store/primary/cache.go

```diff
@@ -0,0 +1,312 @@
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	walkFunc := func(k string, v interface{}) bool {
+		values := v.([]*store.IndexEntry)
+		valCount += len(values)
+		for _, val := range values {
+			unique[val] = struct{}{}
+		}
+		return false
+	}
+
+	cidCount = c.current.Len()
+	c.current.Walk("", walkFunc)
+	if c.previous != nil {
+		cidCount += c.previous.Len()
+		c.previous.Walk("", walkFunc)
+	}
+
+	interned = c.curEnts.Len()
+	if c.prevEnts != nil {
+		interned += c.prevEnts.Len()
+	}
+
+	return CacheStats{
+		Cids:           cidCount,
+		Values:         valCount,
+		UniqueValues:   len(unique),
+		InternedValues: interned,
+		Rotations:      c.rotations,
+	}
+}
+
+func (c *syncCache) get(k string) ([]*store.IndexEntry, bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.getNoLock(k)
+}
+
+func (c *syncCache) getNoLock(k string) ([]*store.IndexEntry, bool) {
+	// Search current cache
+	v, found := c.current.Get(k)
+	if !found {
+		if c.previous == nil {
+			return nil, false
+		}
+
+		// Search previous if not found in current
+		v, found = c.previous.Get(k)
+		if !found {
+			return nil, false
+		}
+
+		// Pull the interned entries for these values forward from the previous
+		// cache to keep using the same pointers as the va used in the cache
+		// values that get pulled forward.
+		values := v.([]*store.IndexEntry)
+		for i, val := range values {
+			values[i] = c.internEntry(val)
+		}
+
+		// Move the value found in the previous tree into the current one.
+		c.current.Put(k, values)
+		c.previous.Delete(k)
+	}
+	return v.([]*store.IndexEntry), true
+}
+
+// put stores an entry in the cache if the entry is not already stored.
+// Returns true if a new entry was added to the cache.
+func (c *syncCache) put(k string, entry store.IndexEntry, rotateSize int) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Get from current or previous cache
+	old, found := c.getNoLock(k)
+	// If found values(s) then check the value to put is already there.
+	if found && duplicateEntry(entry, old) {
+		return false
+	}
+
+	if c.current.Len() >= rotateSize {
+		c.rotate()
+	}
+
+	c.current.Put(k, append(old, c.internEntry(&entry)))
+	return true
+}
+
+func (c *syncCache) putInterned(k string, ent *store.IndexEntry, rotateSize int) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	old, found := c.getNoLock(k)
+	if found && duplicateEntry(*ent, old) {
+		return false
+	}
+
+	if c.current.Len() >= rotateSize {
+		c.rotate()
+	}
+
+	c.current.Put(k, append(old, ent))
+	return true
+}
+
+func (c *syncCache) putMany(keys []string, entry store.IndexEntry, rotateSize int) int {
+	var count int
+	var ent *store.IndexEntry
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	for _, k := range keys {
+		old, found := c.getNoLock(k)
+		if found && duplicateEntry(entry, old) {
+			continue
+		}
+		if ent == nil {
+			ent = c.internEntry(&entry)
+		}
+		c.current.Put(k, append(old, ent))
+		count++
+	}
+
+	if c.current.Len() > rotateSize {
```

Comment:
Shouldn't we check this after every put instead of doing it at the end?

Reply:
Yes... since there is no telling how many they might be adding.

Reply:
Fixed